cpp/inc/bond/core/capped_allocator.h

```diff
@@ -0,0 +1,680 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#pragma once
+
+#include "config.h"
+#include <boost/optional.hpp>
+#include <boost/static_assert.hpp>
+#include <boost/utility/enable_if.hpp>
+#include <boost/core/ref.hpp>
+#include <boost/intrusive_ptr.hpp>
+#include <atomic>
+#include <functional>
+
+
+namespace bond
+{
+    /// @brief Non-sharable counter to be used with \ref capped_allocator.
+    ///
+    /// @tparam T underlying counter type.
+    ///
+    /// @tparam SingleThreaded flag to indicate thread safety.
+    template <typename T = std::size_t, bool SingleThreaded = false>
```

Comment:
What's the relationship between "non-sharable" and SingleThreaded?

Reply:
Unrelated. I should probably describe it as non-copyable. The `SingleThreaded` is only about thread safety.